extensions/redis-cache/runtime/src/main/java/io/quarkus/cache/redis/runtime/RedisCacheImpl.java

```diff
@@ -296,28 +311,36 @@ private void enforceDefaultType() {
         }
     }
 
-    @SuppressWarnings("unchecked")
     @Override
     public <K, V> Uni<V> getOrDefault(K key, V defaultValue) {
         enforceDefaultType();
         byte[] encodedKey = marshaller.encode(computeActualKey(encodeKey(key)));
         return withConnection(new Function<RedisConnection, Uni<V>>() {
             @Override
             public Uni<V> apply(RedisConnection redisConnection) {
-                return (Uni<V>) doGet(redisConnection, encodedKey, classOfValue, marshaller);
+                return doGet(redisConnection, encodedKey, classOfValue, marshaller);
             }
         }).onItem().ifNull().continueWith(new StaticSupplier<>(defaultValue));
     }
 
     @Override
-    @SuppressWarnings("unchecked")
     public <K, V> Uni<V> getOrNull(K key, Class<V> clazz) {
+        return getOrNull(key, (Type) clazz);
+    }
+
+    @Override
+    public <K, V> Uni<V> getOrNull(K key, TypeLiteral<V> type) {
+        return getOrNull(key, type.getType());
+    }
+
+    private <K, V> Uni<V> getOrNull(K key, Type type) {
         enforceDefaultType();
         byte[] encodedKey = marshaller.encode(computeActualKey(encodeKey(key)));
         return withConnection(new Function<RedisConnection, Uni<V>>() {
             @Override
             public Uni<V> apply(RedisConnection redisConnection) {
-                return (Uni<V>) doGet(redisConnection, encodedKey, classOfValue, marshaller);
+                // TODO maybe use `type` (if non-null?) instead of `classOfValue`?
+                return doGet(redisConnection, encodedKey, classOfValue, marshaller);
```

Comment:
Yes, that's a bug 😅

Reply:
I have another commit that makes the `RedisCache` API overall more regular (all `get*` methods have 3 variants: without type, using the default value type, with `Class` type, and with `TypeLiteral` type). I'll submit that in another PR so we can discuss properly, if you agree.